File: codes/python/307.区域和检索-数组可修改.py

class SegmentTreeNode:
    def __init__(self) -> None:
        self.left = -1
        self.right = -1
        self.val = 0


class NumArray:

    def __init__(self, nums: List[int]):
        self.nums = nums
        self.size = len(nums)
        self.tree = [SegmentTreeNode() for _ in range(self.size * 4)]
        self.build(0, 0, self.size - 1)


    def update(self, index: int, val: int) -> None:
        if index < 1:
            return
        self.nums[index] = val
        self.updateRecursion(index, val, 0, 0, self.size - 1)

    def updateRecursion(self, index : int, val : int, nodeIndex : int, nodeLeft : int, nodeRight : int):
        if nodeLeft == nodeRight:
            if nodeIndex == nodeLeft:
                self.tree[nodeIndex].val = val
            return
        mid = nodeLeft + (nodeRight - nodeLeft) // 2
        leftIndex = nodeIndex * 2 + 1
        rightIndex = nodeIndex * 2 + 2
        if index <= mid:
            self.updateRecursion(index, val, leftIndex, nodeLeft, mid)
        else:
            self.updateRecursion(index, val, rightIndex, mid + 1, nodeRight)
        self.tree[nodeIndex].val = self.tree[leftIndex].val + self.tree[rightIndex].val

    def sumRange(self, left: int, right: int) -> int:
        return self.sumRangeRecursion(left, right, 0, 0, self.size - 1)

    def sumRangeRecursion(self, queryLeft : int, queryRight : int, startIndex : int, nodeLeft : int, nodeRight : int):
        if queryLeft > nodeRight or queryRight < nodeLeft:
            return 0
        if queryLeft <= nodeLeft and queryRight >= nodeRight:
            return self.tree[startIndex].val
        mid = nodeLeft + (nodeRight - nodeLeft) // 2
        left_index = startIndex * 2 + 1
        right_index = startIndex * 2 +2
        left_sum_range = self.sumRangeRecursion(queryLeft, queryRight, left_index, nodeLeft, mid)
        right_sum_range = self.sumRangeRecursion(queryLeft, queryRight, right_index, mid + 1, nodeRight)
        return left_sum_range + right_sum_range

    def build(self, index : int, left : int, right : int):
        if left == right:
            self.tree[index].left = left
            self.tree[index].right = right
            self.tree[index].val = self.nums[left]
            return
        mid = left + (right - left) // 2
        leftIndex = index * 2 + 1
        rightIndex = index * 2 + 2
        self.build(leftIndex, left, mid)
        self.build(rightIndex, mid + 1, right)
        self.tree[index].left = left
        self.tree[index].right = right
        self.tree[index].val = self.tree[leftIndex].val + self.tree[rightIndex].val



# Your NumArray object will be instantiated and called as such:
# obj = NumArray(nums)
# obj.update(index,val)
# param_2 = obj.sumRange(left,right)

File: codes/java/307.区域和检索-数组可修改.java

class SegmentTreeNode {
	int left = -1;
	int right = -1;
	int val = 0;
}

class NumArray {

	private int[] nums;
	private int size;
	private SegmentTreeNode[] tree;

	public NumArray(int[] nums) {
		this.nums = nums;
		this.size = nums.length;
		this.tree = new SegmentTreeNode[size * 4];
		build(0, 0, size - 1);
	}

	public void update(int index, int val) {
		if (index < 0) {
			return;
		}
		nums[index] = val;
		updateRecursion(index, val, 0, 0, size - 1);
	}

	private void updateRecursion(int index, int val, int nodeIndex, int nodeLeft, int nodeRight) {
		if (nodeLeft == nodeRight) {
			if (index == nodeLeft) {
				tree[nodeIndex].val = val;
			}
			return;
		}
		int mid = nodeLeft + (nodeRight - nodeLeft) / 2;
		int leftIndex = nodeIndex * 2 + 1;
		int rightIndex = nodeIndex * 2 + 2;
		if (index <= mid) {
			updateRecursion(index, val, leftIndex, nodeLeft, mid);
		} else {
			updateRecursion(index, val, rightIndex, mid + 1, nodeRight);
		}
		tree[nodeIndex].val = tree[leftIndex].val + tree[rightIndex].val;
	}

	public int sumRange(int left, int right) {
		return sumRangeRecursion(left, right, 0, 0, size - 1);
	}

	private int sumRangeRecursion(int queryLeft, int queryRight, int startIndex, int nodeLeft, int nodeRight) {
		if (queryLeft > nodeRight || queryRight < nodeLeft) {
			return 0;
		}
		if (queryLeft <= nodeLeft && queryRight >= nodeRight) {
			return tree[startIndex].val;
		}
		int mid = nodeLeft + (nodeRight - nodeLeft) / 2;
		int leftIndex = startIndex * 2 + 1;
		int rightIndex = startIndex * 2 + 2;
		int leftSumRange = sumRangeRecursion(queryLeft, queryRight, leftIndex, nodeLeft, mid);
		int rightSumRange = sumRangeRecursion(queryLeft, queryRight, rightIndex, mid + 1, nodeRight);
		return leftSumRange + rightSumRange;
	}

	private void build(int index, int left, int right) {
		// 如果left与right的相等，说明区分被划分到只剩一个元素了，那么此时节点的值就是nums[left]
		if (left == right) {
			tree[index] = new SegmentTreeNode();
			tree[index].left = left;
			tree[index].right = right;
			tree[index].val = nums[left];
			return;
		}
		// 否则就继续递归划分区间
		int mid = left + (right - left) / 2;
		int leftIndex = index * 2 + 1;
		int rightIndex = index * 2 + 2;
		build(leftIndex, left, mid);
		build(rightIndex, mid + 1, right);
		// 然后开始计算节点的值
		tree[index] = new SegmentTreeNode();
		tree[index].left = left;
		tree[index].right = right;
		tree[index].val = tree[leftIndex].val + tree[rightIndex].val;
	}
}

/**
 * Your NumArray object will be instantiated and called as such:
 * NumArray obj = new NumArray(nums);
 * obj.update(index,val);
 * int param_2 = obj.sumRange(left,right);
 */